lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp

```diff
@@ -186,95 +151,133 @@ static ArrayAttr replaceArrayAttrElement(ArrayAttr array, size_t elem,
-        IntegerAttr::get(
-            IntegerType::get(ref.op->getContext(), 32, IntegerType::Signless),
-            ref.fieldIdx));
-    annotation = DictionaryAttr::get(ref.op->getContext(), annoField);
+        StringAttr::get(context, "circt.fieldID"),
+        IntegerAttr::get(IntegerType::get(context, 32, IntegerType::Signless),
+                         fieldIdx));
+    annotation = DictionaryAttr::get(context, annoField);
   } else {
-    annotation = DictionaryAttr::get(ref.op->getContext(), anno);
+    annotation = DictionaryAttr::get(context, anno);
   }
 
-  if (!ref.isPort()) {
-    auto newAnno = appendArrayAttr(getAnnotationsFrom(ref.op), annotation);
-    ref.op->setAttr(getAnnotationAttrName(), newAnno);
+  if (ref.isa<OpAnnoTarget>()) {
+    auto newAnno = appendArrayAttr(getAnnotationsFrom(ref.getOp()), annotation);
+    ref.getOp()->setAttr(getAnnotationAttrName(), newAnno);
     return;
   }
 
-  auto portAnnoRaw = ref.op->getAttr(getPortAnnotationAttrName());
+  auto portRef = ref.cast<PortAnnoTarget>();
+  auto portAnnoRaw = ref.getOp()->getAttr(getPortAnnotationAttrName());
   ArrayAttr portAnno = portAnnoRaw.dyn_cast_or_null<ArrayAttr>();
-  if (!portAnno || portAnno.size() != getNumPorts(ref.op)) {
+  if (!portAnno || portAnno.size() != getNumPorts(ref.getOp())) {
     SmallVector<Attribute> emptyPortAttr(
-        getNumPorts(ref.op), ArrayAttr::get(ref.op->getContext(), {}));
-    portAnno = ArrayAttr::get(ref.op->getContext(), emptyPortAttr);
+        getNumPorts(ref.getOp()),
+        ArrayAttr::get(ref.getOp()->getContext(), {}));
+    portAnno = ArrayAttr::get(ref.getOp()->getContext(), emptyPortAttr);
   }
   portAnno = replaceArrayAttrElement(
-      portAnno, ref.portNum,
-      appendArrayAttr(portAnno[ref.portNum].dyn_cast<ArrayAttr>(), annotation));
-  ref.op->setAttr("portAnnotations", portAnno);
+      portAnno, portRef.getPortNo(),
+      appendArrayAttr(portAnno[portRef.getPortNo()].dyn_cast<ArrayAttr>(),
+                      annotation));
+  ref.getOp()->setAttr("portAnnotations", portAnno);
 }
 
 // Some types have been expanded so the first layer of aggregate path is
 // a return value.
-static LogicalResult updateExpandedPort(StringRef field, AnnoTarget &entity) {
-  if (auto mem = dyn_cast<MemOp>(entity.op))
+static LogicalResult updateExpandedPort(StringRef field, AnnoTarget &ref) {
+  if (auto mem = dyn_cast<MemOp>(ref.getOp()))
     for (size_t p = 0, pe = mem.portNames().size(); p < pe; ++p)
       if (mem.getPortNameStr(p) == field) {
-        entity.portNum = p;
+        ref = PortAnnoTarget(ref.getOp(), p);
         return success();
       }
-  if (auto inst = dyn_cast<InstanceOp>(entity.op))
+  if (auto inst = dyn_cast<InstanceOp>(ref.getOp()))
     for (size_t p = 0, pe = inst.getNumResults(); p < pe; ++p)
       if (inst.getPortNameStr(p) == field) {
-        entity.portNum = p;
+        ref = PortAnnoTarget(ref.getOp(), p);
```

Comment:
This part here is sketchy.  From what we were saying an instance op shouldn't get any annotations.  You shouldn't even be able to convert a instance op into a `PortAnnoTarget` but I guess `Operation *` is implicitly convertible to `FModuleLike`.  I'll see if I can fix this up to put the annotations on the module without this exploding.

Reply:
This used to work correctly until I started splitting the API here https://github.com/llvm/circt/pull/2510#discussion_r792008401 and I missed this problem until right now.

Reply:
The issue I think this was trying to address is that the annotations assume that instances and memories and stuff return a single bundle, so an annotation could point to a field of that bundle.  But since the parser is breaking bundles into multiple return values, the outer layer of naming has to dig through port names.

Reply:
Specifically what was happening is the `Operation *` (which was an `InstanceOp`) was being converted to a `FModuleLike` and it was asserting after trying to find the interface.  When I first wrote `PortAnnoTarget`it used to take an `Operation *` instead of a `FModuleLIke` argument, since I thought we wanted to be able to target the ports on an instance op.